ksqldb-rest-app/src/main/java/io/confluent/ksql/rest/server/CommandTopicMigrationUtil.java

```diff
@@ -0,0 +1,126 @@
+import io.confluent.ksql.rest.server.computation.Command;
+import io.confluent.ksql.rest.server.computation.CommandStore;
+import io.confluent.ksql.rest.server.computation.InternalTopicSerdes;
+import io.confluent.ksql.rest.server.computation.QueuedCommand;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.clients.producer.KafkaProducer;
+import org.apache.kafka.clients.producer.Producer;
+import org.apache.kafka.clients.producer.ProducerConfig;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.serialization.ByteArrayDeserializer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class CommandTopicMigrationUtil {
+
+  private static final Logger log = LoggerFactory.getLogger(CommandTopicMigrationUtil.class);
+
+  private CommandTopicMigrationUtil() {
+
+  }
+
+  public static void commandTopicMigration(
+      final String commandTopic,
+      final KsqlRestConfig restConfig,
+      final KsqlConfig config
+  ) {
+    final TopicPartition topicPartition = new TopicPartition(commandTopic, 0);
+
+    // produce a higher version command to the old command topic so that other servers
+    // stop writing to it since they'll be degraded
+    final Producer<CommandId, Command> oldBrokerProducer = new KafkaProducer<>(
+        config.originals(),
+        InternalTopicSerdes.serializer(),
+        InternalTopicSerdes.serializer()
+    );
+    final ProducerRecord<CommandId, Command> degradedCommand = new ProducerRecord<>(
+        commandTopic,
+        topicPartition.partition(),
+        new CommandId(CommandId.Type.CLUSTER, "", CommandId.Action.ALTER),
+        new Command(
+            "",
+            Collections.emptyMap(),
+            Collections.emptyMap(),
+            Optional.empty(),
+            Optional.of(Integer.MAX_VALUE - 1),
+            Integer.MAX_VALUE
+        )
+    );
+    oldBrokerProducer.send(degradedCommand);
+    oldBrokerProducer.close();
+
+    // read all the commands from the old command topic
+    final org.apache.kafka.clients.consumer.Consumer<byte[], byte[]> oldBrokerConsumer =
+        new KafkaConsumer<>(
+            config.originals(),
+            new ByteArrayDeserializer(),
+            new ByteArrayDeserializer()
+    );
+    oldBrokerConsumer.assign(Collections.singleton(topicPartition));
+
+    final List<QueuedCommand> commands = CommandTopic.getAllCommandsInCommandTopic(
+        oldBrokerConsumer,
+        topicPartition,
+        Optional.empty(),
+        CommandStore.POLLING_TIMEOUT_FOR_COMMAND_TOPIC
+    );
+    oldBrokerConsumer.close();
+
+    // remove the incompatible command that was just written to the command topic
+    commands.remove(commands.size() - 1);
```

Comment:
Add an assertion here that the command that was removed was the bad command, or better yet find the bad command in the list and get the prefix that excludes it. It actually may not be the last command for multi-node clusters

Reply:
OK, I made the commandId a bit more unique that triggers the degraded state in the other servers so that we can easily identify it when dropping it here. The main benefit I think is that we may produce multiple of these bad commands to the old command topic when attempting migration if for some reason the MIGRATOR fails one attempt, the server restarts and tries the migration again. We'd be able to drop all of the migration commands.

>It actually may not be the last command for multi-node clusters
What situation would cause this? Wouldn't any ongoing transactions fail when the migration command gets produced? And then the other servers would have to process the migration command before enqueueing any other command.